### PR TITLE
Fix bug in redirection of unit test outputs

### PR DESCRIPTION
**Pull request by gufe44**
Otherwise it builds and tests fine on OpenBSD, builds fine on NetBSD but (edit: fixed in #3540) gets stuck in mbedtls_mpi_gen_prime (corner case limb size) of test_suite_mpi (both stable/x86_64).

### CODE REVIEW

**Comment by danh-arm** [comment]
The test code does not build with this PR. From the Travis logs:
```
******************************************************************

* test_default_out_of_box: test: main suites make, default config (out-of-box) 

* Thu Jul 30 07:35:27 UTC 2020

******************************************************************

test_suite_aes.cbc ................................................ Use of uninitialized value $skipped in subtraction (-) at scripts/run-test-suites.pl line 127.

Use of uninitialized value $tests in subtraction (-) at scripts/run-test-suites.pl line 127.

FAIL

test_suite_aes.cfb ................................................ Use of uninitialized value $skipped in subtraction (-) at scripts/run-test-suites.pl line 127.

Use of uninitialized value $tests in subtraction (-) at scripts/run-test-suites.pl line 127.

FAIL
...
```

**Comment by danh-arm** [comment]
Also needs:
* A signed-off-by line
* A commit message with rationale as to what the problem is why this fixes it (more than just "my config works with this").
* A change log entry.

**Comment by gilles-peskine-arm** [comment]
This was previously reported in https://github.com/ARMmbed/mbedtls/issues/2311. There's also a proposed patch in that report, but please note that **we cannot legally accept that patch, or a patch derived from it, without an explicit permission grant from its author**, so you need to write your patch independently. We also haven't confirmed that this patch works.

I would prefer a different approach. Instead of modifying `stdout`, change references to `stdout` in `host_test.function` to `verbose_out` and set `verbose_out` and
```
FILE *verbose_out = stdout;
#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
if (option_verbose) { ... modify verbose_out ... }
#endif
```
And close `verbose_out` before exiting if it isn't stdout, so that tools such as ASan don't complain of a memory leak.

**Comment by gufe44** [comment]
@gilles-peskine-arm, I'm very sorry, can you please clarify why the changes I made modify the behaviour of the program at all? Wouldn't your proposed solution need the tests themselves also to reference verbose_out?

**Comment by gilles-peskine-arm** [comment]
The test code never references `stdout` explicitly, or a function that has stdout hard-coded such as `printf`. It has to work on embedded targets that don't have stdio. Only `host_test.function` references stdout (and `helpers.function`, but that code is specific to hosted targets so it should actually have been in `host_test.function`).

I haven't looked at your patch too closely. It clearly does modify the behavior of the program since `make test` isn't working anymore on the CI.

**Comment by gufe44** [comment]
I must have been blind. Would something like this be portable enough?

**Comment by gufe44** [comment]
Does this need more work? @danh-arm @gilles-peskine-arm

File: ChangeLog.d/stdout-macro.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Bugfix
+   * Duplicate file descriptor into STDOUT_FILENO instead of replacing stdout
```
Comment (hanno-arm) [inline ChangeLog.d/stdout-macro.txt:2]: Thanks for adding the ChangeLog entry. Generally, it is more valuable to users to describe context and symptoms of the problem rather than the mechanics of the fix. Suggestion:
```
    * Fix bug in redirection of unit test outputs which previously lead to some tests 
      getting stuck on NetBSD. Reported in #2311 and fix contributed in #3528 by XXX.
```
Feel free to be more specific and to add your name here as well, if you wish.

File: tests/suites/helpers.function [hunk 2]
```diff
@@ -446,44 +446,51 @@ void mbedtls_param_failed( const char *failure_condition,
...
     {
-        close( stdout_fd );
+        close( dup_fd );
+        fclose( path_stream );
         return -1;
     }
 
-    return stdout_fd;
+    fclose( path_stream );
+    return dup_fd;
```
Comment (hanno-arm) [inline tests/suites/helpers.function:478]:
```suggestion
    return( dup_fd );
```

File: tests/suites/helpers.function [hunk 3]
```diff
@@ -446,44 +446,51 @@ void mbedtls_param_failed( const char *failure_condition,
...
+    }
 
-    if( *out_stream == NULL )
+    fflush( out_stream );
+    if( dup2( fileno( path_stream ), out_fd ) == -1)
     {
-        close( stdout_fd );
+        close( dup_fd );
+        fclose( path_stream );
         return -1;
```
Comment (hanno-arm) [inline tests/suites/helpers.function:474]:
```suggestion
        return( -1 );
```
I realize the pre-existing code didn't obey the coding style either...

File: tests/suites/helpers.function [hunk 4]
```diff
@@ -446,44 +446,51 @@ void mbedtls_param_failed( const char *failure_condition,
...
+    int out_fd, dup_fd;
+    FILE* path_stream;
 
-    if( stdout_fd == -1 )
+    out_fd = fileno( out_stream );
+    dup_fd = dup( out_fd );
+
+    if( dup_fd == -1 )
     {
         return -1;
```
Comment (hanno-arm) [inline tests/suites/helpers.function:459]:
```suggestion
        return( -1 );
```

File: tests/suites/helpers.function [hunk 5]
```diff
@@ -446,44 +446,51 @@ void mbedtls_param_failed( const char *failure_condition,
...
     }
 
-    fflush( *out_stream );
-    fclose( *out_stream );
-    *out_stream = fopen( path, "w" );
+    path_stream = fopen( path, "w" );
+    if( path_stream == NULL )
+    {
+        close( dup_fd );
+        return -1;
```
Comment (hanno-arm) [inline tests/suites/helpers.function:466]:
```suggestion
        return( -1 );
```

**Comment by hanno-arm** [comment]
@gufe44 `make test` is failing in the full configuration (i.e. after running `./scripts/config.py full`). Could you investigate?

**Comment by gufe44** [comment]
@hanno-arm Isn't this a problem on the development branch? I have only looked so far. I think the tests making use of psa_crypto.c fail to gather entropy with mbedtls_entropy_func. Tests pass if test_suite_entropy has been run before and tests/seedfile was created (or if make test is run twice).

**Comment by gilles-peskine-arm** [comment]
@gufe44 @hanno-arm Indeed there's a gotcha in our test suites: if `MBEDTLS_ENTROPY_NV_SEED` is enabled (which is not the case by default but is the case in the `full` config), the unit tests that call the entropy module need an existing `seedfile`. `test_suite_entropy` is an exception because it contains a test case for the creation of a seedfile that runs before any test case that needs a seedfile.

Historically this has mostly worked invisibly because the other test suites that call the entropy module come after `entropy` in alphabetical order. But if you enable `MBEDTLS_USE_PSA_CRYPTO` (which `config.py full` does) then `test_suite_cipher` calls `psa_crypto_init` which obtains entropy and therefore requires a seedfile. Since `cipher` comes before `entropy`, if you haven't created a seedfile manually, `psa_crypto_init()` fails.

Another failure reason is when the seedfile is too small. The size of the seedfile is determined by 
the hash that the entropy module uses. If you run tests with either `MBEDTLS_SHA512_C` disabled or `MBEDTLS_ENTROPY_FORCE_SHA256` enabled, and then run tests with `MBEDTLS_SHA512_C` enabled or `MBEDTLS_ENTROPY_FORCE_SHA256` disabled, and `MBEDTLS_ENTROPY_NV_SEED` enabled in both cases, the second test run starts with a seedfile that's too small, and again it will fail if something needs entropy before `test_suite_entropy`.

Enough people have gotten bitten by this that we should fix it rather than document it. https://github.com/ARMmbed/mbedtls/issues/3575

File: tests/suites/helpers.function [hunk 6]
```diff
@@ -446,44 +446,51 @@ void mbedtls_param_failed( const char *failure_condition,
...
+    path_stream = fopen( path, "w" );
+    if( path_stream == NULL )
+    {
+        close( dup_fd );
+        return( -1 );
+    }
 
-    if( *out_stream == NULL )
+    fflush( out_stream );
+    if( dup2( fileno( path_stream ), out_fd ) == -1)
```
Comment (hanno-arm) [inline tests/suites/helpers.function:470]: Trivial style issue: 
```suggestion
    if( dup2( fileno( path_stream ), out_fd ) == -1 )
```

**Review by hanno-arm — APPROVED** [review]
Thank you for your contribution, the patch looks good to me.

There's a trivial style issue remaining but it doesn't block the PR. Feel free to push a commit or leave things as they are.

Apologies for mistakenly blaming your PR for one of the CI failures. As far as I see, they're known issues in the CI, but I'd be grateful if @mpg or @gilles-peskine-arm could double-check and confirm for the latest run, as they're more familiar with this these days.

**Comment by gufe44** [comment]
Thank you for your review and for pointing out the remaining style issue. Are there any concerns about portability or backporting to LTS branches?

**Review by hanno-arm — APPROVED** [review]

**Comment by hanno-arm** [comment]
@gufe44 Have you checked whether the underlying issue is present in 2.7 and 2.16?

**Comment by gufe44** [comment]
Yes. The relevant functions were already present mostly in the same files.
